### DISCUSSION TITLE
Selections as boolean arrays

### DISCUSSION THREAD

**Discussion by BradyAJohnston** [MDAnalysis/mdanalysis#4555] [General]
Would it be possible to get the result of an atom group selection as a boolean array for the parent atom group?

I am getting it as an array as follows, but I'd like to potentially remove the need for the `np.isin()` to speed it up.
```py
def bool_selection(ag, selection, **kwargs) -> np.ndarray:
    return np.isin(ag.ix, ag.select_atoms(selection, **kwargs).ix).astype(bool)
```
I have a bit of a peak at the source code and it seems like it is likely not possible, but I'm also not very familiar with how it all works.

I'm using it for masking the parent atom group inside of Blender for updating selections, as updating the actual atom group that is associated with the object would be quite difficult.

https://github.com/MDAnalysis/mdanalysis/assets/36021261/8e8afe26-837a-4aa9-98ee-55592fabaa70

**Comment by orbeckst** [comment]
I think what you have is as close as it gets. The selection mechanism is pretty closely tied to generating AtomGroups, but given that an AtomGroup is effectively just the index array (ix), we generally don't consider it much of a draw back. 

However, `select_atoms()` is an expensive operation so you really don't want to do it repeatedly. A common fallacy is to do a static selection inside a loop instead of using the AtomGroup, as in
```python
# BAD CODE
u = mda.Universe(TOP, TRJ)
results = []
for ts in u.trajectory:
   p = u.select_atoms("protein")   # bad! creates same AG repeatedly
   results.append(p.radius_of_gyration())
```
Instead
```python
# BETTER
u = mda.Universe(TOP, TRJ)
results = []
p = u.select_atoms("protein")   # only make static selection once
for ts in u.trajectory:
   results.append(p.radius_of_gyration())
```

The alternative is to completely circumvent the selection parser and roll your own selections as in
```python
ca = u.atoms[u.atoms.names == "CA"]
```
which works well for many simple selections but is not quite as general for the typical user. Of course, in your case `u.atoms.names == "CA"` would already be the boolean array you want.

**Reply by orbeckst** [reply]
Not sure who poked at selections recently... maybe @lilyminium or @richardjgowers may have some insights?

**Reply by orbeckst** [reply]
> I'm using it for masking the parent atom group inside of Blender for **updating selections**, as updating the actual atom group that is associated with the object would be quite difficult.

Only noticed now that you need *updating selections* — then indeed you have to effectively run the selection *inside the loop* and that's going to be expensive no matter what. I doubt that `np.isin` is the performance bottleneck (if this what you're concerned about).

**Reply by BradyAJohnston** [reply]
That was maybe going to be a follow up question. Currently in the loop I am creating the selection anew each time (which yes is expensive). If I could get a selection that updated output a boolean array that would be amazing, but it's looking like I might just need to create a new AtomGroup from the selection that can update & store that somewehere to track, then each frame just compare what is in that AtomGroup to the parent Universe and get the boolean mask from there.

**Reply by BradyAJohnston** [reply]
Appreciate the info and the examples as well. You can already create selections inside of the nodes themselves, but this is just an attempt to give the 'familiar' interface of MDA selection language to the user, so I'll have to stick with the selection parser. I should be able to have a check when I am changing frames to see if the selection has updated and if it has create a new AtomGroup, and if not then just use the AtomGroup that exists / is already updating and pull the `.ix` from that to get the mask.
